### ISSUE TITLE
Cannot decorate CoreShop\Bundle\ResourceBundle\Pimcore\Repository\StackRepository

### ISSUE THREAD

**Issue by dkarlovi** [coreshop/CoreShop#2194]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no

Several places in the codebase expect specific `StackRepository` injected, example:

https://github.com/coreshop/CoreShop/blob/9a18128ca22c0b92256be645a5cb443db7df1760/src/CoreShop/Component/Core/Cart/Rule/Action/GiftProductActionProcessor.php#L35

https://github.com/coreshop/CoreShop/blob/9a18128ca22c0b92256be645a5cb443db7df1760/src/CoreShop/Bundle/CoreBundle/Controller/ProductValidationController.php#L99

https://github.com/coreshop/CoreShop/blob/9a18128ca22c0b92256be645a5cb443db7df1760/src/CoreShop/Bundle/CoreBundle/Report/ProductsReport.php#L44

https://github.com/coreshop/CoreShop/blob/9a18128ca22c0b92256be645a5cb443db7df1760/src/CoreShop/Bundle/ProductBundle/Controller/ProductUnitDefinitionsController.php#L38

https://github.com/coreshop/CoreShop/blob/9a18128ca22c0b92256be645a5cb443db7df1760/src/CoreShop/Bundle/ProductBundle/Controller/ProductUnitDefinitionsController.php#L55

**Comment by dpfaffenbauer** [comment]
are you trying to decorate it?

**Comment by dkarlovi** [comment]
Correct, we need to be able to do some work after the product is pulled out of it.

**Comment by dpfaffenbauer** [comment]
how do you decorate it? a bit more context would be heplful ;)

**Comment by dkarlovi** [comment]
Basically, we're decorating it to create a proxy objects of the product(s) it finds, something like

```php
    public function forceFind($id, bool $force = true): ?object
    {
        $object = $this->repository->forceFind($id, $force);

        return $this->createProxy($object);
    }
    ```
    
Config looks something like

```yaml

    sigwin_pimcore_elasticsearch.repository.wishlist:
        class: Sigwin\PimcoreElasticsearchBundle\Repository\ProxyStackRepository
        decorates: coreshop.repository.stack.wishlist_product
        arguments:
            - '@.inner'
            - '@sigwin_pimcore_elasticsearch.factory.proxy_class'
            - '@sigwin_pimcore_elasticsearch.context_checker.cart'
            - '@sigwin_pimcore_elasticsearch.context_checker.wishlist'

    sigwin_pimcore_elasticsearch.repository.product:
        class: Sigwin\PimcoreElasticsearchBundle\Repository\ProxyStackRepository
        decorates: coreshop.repository.stack.purchasable
        arguments:
            - '@.inner'
            - '@sigwin_pimcore_elasticsearch.factory.proxy_class'
            - '@sigwin_pimcore_elasticsearch.context_checker.cart'
            - '@sigwin_pimcore_elasticsearch.context_checker.wishlist'
```

**Comment by dkarlovi** [comment]
Fixed in #2196.
